libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1759,22 +1817,35 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
 			leftType->humanReadableName() +
 			" and " +
 			rightType->humanReadableName() +
-			(!result.message().empty() ? ". " + result.message() : "")
+			"." +
+			(!builtinResult.message().empty() ? " " + builtinResult.message() : "") +
+			(!userDefinedOperatorResult.message().empty() ? " " + userDefinedOperatorResult.message() : "")
+		);
+
+	if (builtinResult)
+		commonType = builtinResult.get();
+	else if (userDefinedOperatorResult)
+	{
+		solAssert(
+			userDefinedFunctionType->parameterTypes().size() == 2 &&
+			*userDefinedFunctionType->parameterTypes().at(0) ==
+			*userDefinedFunctionType->parameterTypes().at(1)
 		);
-		commonType = leftType;
+		commonType = userDefinedFunctionType->returnParameterTypes().at(0);
 	}
+
 	_operation.annotation().commonType = commonType;
 	_operation.annotation().type =
 		TokenTraits::isCompareOp(_operation.getOperator()) ?
 		TypeProvider::boolean() :
 		commonType;
-	_operation.annotation().isPure =
-		*_operation.leftExpression().annotation().isPure &&
-		*_operation.rightExpression().annotation().isPure;
-	_operation.annotation().isLValue = false;
-	_operation.annotation().isConstant = false;
 
-	if (_operation.getOperator() == Token::Exp || _operation.getOperator() == Token::SHL)
+	if (userDefinedOperatorResult)
+		solAssert(
```

Comment:
I added a description to the assertion and new test cases. Additionally, I decided to update `Type::userDefinedOperator` so it includes return parameters during user-defined operator search. As a result of this we avoid ICE in case of invalid return parameters in the function attached to the operator.